File: src/main/java/com/plugin/frege/psi/FregePsiElementFinder.java
package com.plugin.frege.psi;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.ProjectFileIndex;
import com.intellij.psi.*;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class FregePsiElementFinder extends PsiElementFinder {
    @Override
    public @Nullable PsiClass findClass(@NotNull String qualifiedName, @NotNull GlobalSearchScope scope) {
        PsiClass[] classes = findClasses(qualifiedName, scope);
        if (classes.length == 0) {
            return null;
        }

        return classes[0];
    }

    @Override
    public PsiClass @NotNull [] findClasses(@NotNull String qualifiedName, @NotNull GlobalSearchScope scope) {
        return getClasses(clazz -> Objects.equals(clazz.getQualifiedName(), qualifiedName), scope);
    }

    @Override
    public PsiClass @NotNull [] getClasses(@NotNull PsiPackage psiPackage, @NotNull GlobalSearchScope scope) {
        return getClasses(clazz -> {
            String clazzName = clazz.getName();
            return clazzName != null && psiPackage.containsClassNamed(clazzName);
        }, scope);
    }

    private PsiClass @NotNull [] getClasses(@NotNull Predicate<PsiClass> predicate, @NotNull GlobalSearchScope scope) {
        Project project = scope.getProject();
        if (project == null) {
            return PsiClass.EMPTY_ARRAY;
        }

        PsiManager manager = PsiManager.getInstance(project);
        List<PsiClass> classes = new ArrayList<>();

        ProjectFileIndex.SERVICE.getInstance(project).iterateContent(fileOrDir -> {
            if (fileOrDir.isDirectory() || !scope.contains(fileOrDir)) {
                return true;
            }

            PsiFile file = manager.findFile(fileOrDir);
            if (!(file instanceof FregeFile)) {
                return true;
            }

            classes.addAll(PsiTreeUtil.findChildrenOfType(file, FregePsiClass.class).stream()
                    .filter(predicate).collect(Collectors.toList()));

            return true;
        });

        return classes.toArray(PsiClass[]::new);
    }
}

File: src/main/kotlin/com/plugin/frege/psi/FregePsiElementFinder.kt
package com.plugin.frege.psi

import com.intellij.openapi.roots.ContentIterator
import com.intellij.openapi.vfs.VirtualFileFilter
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiElementFinder
import com.intellij.psi.PsiManager
import com.intellij.psi.PsiPackage
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.util.PsiTreeUtil
import com.plugin.frege.resolve.FregeResolveUtil
import com.plugin.frege.stubs.index.FregeClassNameIndex

class FregePsiElementFinder : PsiElementFinder() {
    override fun findClass(qualifiedName: String, scope: GlobalSearchScope): PsiClass? {
        val classes = findClasses(qualifiedName, scope)
        return classes.firstOrNull()
    }

    override fun findClasses(qualifiedName: String, scope: GlobalSearchScope): Array<PsiClass> {
        val project = scope.project ?: return PsiClass.EMPTY_ARRAY
        return FregeClassNameIndex.INSTANCE.findByName(qualifiedName, project, scope).toTypedArray()
    }

    override fun getClasses(psiPackage: PsiPackage, scope: GlobalSearchScope): Array<PsiClass> { // TODO improve
        return getClasses({ clazz ->
            val clazzName = clazz.name
            clazzName != null && psiPackage.containsClassNamed(clazzName)
        }, scope)
    }

    private fun getClasses(predicate: (clazz: PsiClass) -> Boolean, scope: GlobalSearchScope): Array<PsiClass> {
        val project = scope.project ?: return PsiClass.EMPTY_ARRAY
        val manager = PsiManager.getInstance(project)
        val classes: MutableList<PsiClass> = ArrayList()

        val processor = ContentIterator { virtualFile ->
            val file = manager.findFile(virtualFile)
            PsiTreeUtil.findChildrenOfType(file, FregePsiClass::class.java)
                .filter(predicate).forEach { classes.add(it) }
            true
        }
        val filter = VirtualFileFilter { true }

        FregeResolveUtil.iterateFregeFiles(processor, scope, filter)
        return classes.toTypedArray()
    }
}
